### ISSUE TITLE
Server context not transferred (`createServerContext`)

### ISSUE THREAD

**Issue by amannn** [vercel/next.js#44222]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:22 PDT 2022; root:xnu-8020.121.3~4/RELEASE_X86_64
    Binaries:
      Node: 16.16.0
      npm: 8.11.0
      Yarn: 1.22.18
      pnpm: 7.1.7
    Relevant packages:
      next: 13.0.8-canary.2
      eslint-config-next: 13.0.7
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue

https://github.com/amannn/nextjs-server-context

### To Reproduce

1. Clone [the repo](https://github.com/amannn/nextjs-server-context)
2. Run `yarn && yarn dev`
3. Go to [http://localhost:3000/](http://localhost:3000/)
4. The server context value is displayed as "Updated server context value (correct)"
5. Navigate to another page
6. The server context value is displayed as "Initial server context value (incorrect)"

Interestingly, [a deployed version of this app](https://nextjs-server-context-jl0f9w3s2-amann.vercel.app/) doesn't have the issue. Maybe this is because all pages are statically generated at build time, where the root layout is used for all pages?

### Describe the Bug

My apologies if it's too early to report bugs for this feature, I saw [this tweet](https://twitter.com/sebmarkbage/status/1587615489605369861) and wanted to experiment with this new API.

I noticed that the value is lost when navigating to other pages.

### Expected Behavior

My understanding is that the context value should be present for all pages, since it should be sent back to the server when another page is requested and the root layout doesn't need to re-render.

### Which browser are you using? (if relevant)

Version 108.0.5359.124 (Official Build) (x86_64)

**Comment by amannn** [comment]
On a related note, if I make use of `use` in an async component, this throws:

<img width="953" alt="Screenshot 2022-12-21 at 11 38 27" src="https://user-images.githubusercontent.com/4038316/208885627-c1bfaeb7-2440-4cc4-831e-0f6844eb8624.png">

I understand that `use` is primarily intended for the client-side, but for APIs that can be used both on the client- as well as the server side, it would be helpful if this wouldn't break the app.

**Comment by hansottowirtz** [comment]
Also running into this issue, so we pass our locale as a prop to our server components for now.

**Comment by amannn** [comment]
Just updated to Next.js 13.4, unfortunately this is still an issue.

**Comment by timneutkens** [comment]
> My apologies if it's too early to report bugs for this feature

Yeah, this is not implemented currently so it's expected to not work well.

**Comment by Drag13** [comment]
> Closed as completed 2 weeks ago. 

I am playing with v. 13.4.3 and this issue is still here. Server context becomes null after changing the route. What did I miss?
